[WebGPU] Validate buffer length for bind group entries

tadeuzagallo opened the pull request:
#### ae36fffb790d1c33f548681ce46b41c007850418
<pre>
[WebGPU] Validate buffer length for bind group entries
<a href="https://bugs.webkit.org/show_bug.cgi?id=273151">https://bugs.webkit.org/show_bug.cgi?id=273151</a>
<a href="https://rdar.apple.com/126848520">rdar://126848520</a>

Reviewed by Mike Wyrzykowski.

We were missing a validation that the size of the buffer passed into a bind group
was big enough to fit the type of that entry. Most of the plumbing was already done,
but we never populated the `minBindingSize` in the compiler info. That works out of
the box for auto layouts, but for user-provided layouts we need to update the entries
with the computed size, and this later needs to be checked from draw/dispatch.

* LayoutTests/fast/webgpu/fuzz-126848520-auto-layout-expected.txt: Added.
* LayoutTests/fast/webgpu/fuzz-126848520-auto-layout.html: Added.
* LayoutTests/fast/webgpu/fuzz-126848520-explicit-layout-expected.txt: Added.
* LayoutTests/fast/webgpu/fuzz-126848520-explicit-layout.html: Added.
* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::rewriteEntryPoints):
* Source/WebGPU/WGSL/EntryPointRewriter.h:
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::RewriteGlobalVariables):
(WGSL::RewriteGlobalVariables::visitEntryPoint):
(WGSL::bindingMemberForGlobal):
(WGSL::RewriteGlobalVariables::insertStructs):
(WGSL::rewriteGlobalVariables):
* Source/WebGPU/WGSL/GlobalVariableRewriter.h:
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepareImpl):
(WGSL::prepare):
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WGSL/wgslc.cpp:
(runWGSL):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::createArgumentDescriptor):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::executePreDispatchCommands):
* Source/WebGPU/WebGPU/Pipeline.h:
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::createLibrary):
(WebGPU::validateBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::executePreDrawCommands):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::earlyCompileShaderModule):

Canonical link: <a href="https://commits.webkit.org/278052@main">https://commits.webkit.org/278052@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7476298cff6a502beb536f0bee003e08bf2e1b14

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49331 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28613 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52364 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34628 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26233 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/52202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51431 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26165 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42491 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23619 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43674 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7699 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45530 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44178 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54082 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24412 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20601 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47625 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25684 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42698 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46618 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10843 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26523 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25408 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (commented):
Seems to crash running:
https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,capability_checks,features,texture_formats:depth_stencil_state:isAsync=false;format=%22depth32float-stencil8%22;enable_required_feature=true

mwyrzykowski commented on Source/WebGPU/WebGPU/Pipeline.mm:
```diff
@@ -55,7 +55,7 @@
         wgslPipelineLayout = ShaderModule::convertPipelineLayout(*pipelineLayout);
 
-    auto prepareResult = WGSL::prepare(*ast, entryPoint, wgslPipelineLayout);
+    auto prepareResult = WGSL::prepare(*ast, entryPoint, &*wgslPipelineLayout);
```
when I run https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,capability_checks,features,texture_formats:depth_stencil_state:isAsync=false;format=%22depth32float-stencil8%22;enable_required_feature=true I see:
```
(lldb) po wgslPipelineLayout
 Has Value=false 
```

mwyrzykowski reviewed (changes requested):
I think this breaks https://webgpu.github.io/webgpu-samples/?sample=deferredRendering

tadeuzagallo committed:
[WebGPU] Validate buffer length for bind group entries
https://bugs.webkit.org/show_bug.cgi?id=273151
rdar://126848520

Reviewed by Mike Wyrzykowski.

We were missing a validation that the size of the buffer passed into a bind group
was big enough to fit the type of that entry. Most of the plumbing was already done,
but we never populated the `minBindingSize` in the compiler info. That works out of
the box for auto layouts, but for user-provided layouts we need to update the entries
with the computed size, and this later needs to be checked from draw/dispatch.

* LayoutTests/fast/webgpu/fuzz-126848520-auto-layout-expected.txt: Added.
* LayoutTests/fast/webgpu/fuzz-126848520-auto-layout.html: Added.
* LayoutTests/fast/webgpu/fuzz-126848520-explicit-layout-expected.txt: Added.
* LayoutTests/fast/webgpu/fuzz-126848520-explicit-layout.html: Added.
* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::rewriteEntryPoints):
* Source/WebGPU/WGSL/EntryPointRewriter.h:
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::RewriteGlobalVariables):
(WGSL::RewriteGlobalVariables::visitEntryPoint):
(WGSL::bindingMemberForGlobal):
(WGSL::RewriteGlobalVariables::insertStructs):
(WGSL::rewriteGlobalVariables):
* Source/WebGPU/WGSL/GlobalVariableRewriter.h:
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepareImpl):
(WGSL::prepare):
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WGSL/wgslc.cpp:
(runWGSL):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::createArgumentDescriptor):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::executePreDispatchCommands):
* Source/WebGPU/WebGPU/Pipeline.h:
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::createLibrary):
(WebGPU::validateBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::executePreDrawCommands):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::earlyCompileShaderModule):

Canonical link: https://commits.webkit.org/278052@main
